Migration to Identity: Issues in the code snippets

sadasant opened the pull request:
This PR fixes two issues in the code snippets:

- We passed the `tokenAudience` to the `ClientSecretCredential` as a parameter when `ClientSecretCredential` does not receive such parameter.
- We were not calling the `getToken` method of the credential used in the sample for the `Regarding scopes` section. We were instead passing it to a credential.
- In the `Regarding scopes` code snippet, we were also moving from `interactiveLogin` to `AzureCliCredential`, which is confusing. Using `InteractiveBrowserCredential` instead.

sadasant committed:
Migration to Identity: Issues in the code snippets

This PR fixes two issues in the code snippets:

- We passed the `tokenAudience` to the `ClientSecretCredential` as a parameter when `ClientSecretCredential` does not receive such parameter.
- We were not calling the `getToken` method of the credential used in the sample for the `Regarding scopes` section. We were instead passing it to a credential.
- In the `Regarding scopes` code snippet, we were also moving from `interactiveLogin` to `AzureCliCredential`, which is confusing. Using `InteractiveBrowserCredential` instead.

ramya-rao-a commented on migrate-to-identity-v2.md:
```diff
@@ -127,7 +127,7 @@ const tokenAudience = "https://graph.microsoft.com/";
 
 async function main() {
 - const credential = new ApplicationTokenCredentials(clientId, domain, secret, tokenAudience, environment);
```
The token audience is set for graph. Why are we passing that when creating the credential from ms-rest-nodeauth in the first place if the sample is about creating a credential to be passed to SubscriptionClient?

sadasant replied:
how does it look now? I’m using the china cloud as the audience.

ramya-rao-a replied:
Can we test it out?

sadasant replied:
sure can! (tomorrow, but will do)

sadasant replied:
This works for me if I use the AzureCloud instead of the China Cloud, but it works 🤔

Here’s my code:

```ts
const tokenAudience = "https://management.azure.com/";
const environment = Environment.AzureCloud;
const clientId = process.env.AZURE_CLIENT_ID;
const domain = process.env.AZURE_TENANT_ID;
const secret = process.env.AZURE_CLIENT_SECRET;
const credential = new ApplicationTokenCredentials(clientId, domain, secret, tokenAudience, environment);
console.log(await credential.getToken());
```

ramya-rao-a replied:
This code snippet we are editing is about getting subscriptions from the SubscriptionClient while your code snippet above gets the token...

sadasant replied:
This works:

```ts
const tokenAudience = "https://management.azure.com/";
const environment = Environment.AzureCloud;
const clientId = process.env.AZURE_CLIENT_ID;
const domain = process.env.AZURE_TENANT_ID;
const secret = process.env.AZURE_CLIENT_SECRET;
const credential = new ApplicationTokenCredentials(clientId, domain, secret, tokenAudience, environment);
const client = new SubscriptionClient(credential, {
  baseUri: environment.resourceManagerEndpointUrl,
  baseUri: "https://management.azure.com/"
});
const subscriptions = await client.subscriptions.list();
```

Which I was expecting, since I was able to get a token in the previous snippet I tried 🤔 Is there something specific that I’m missing?

ramya-rao-a replied:
I am still slightly confused...

Take the below facts:
- `tokenAudience` is used to determine which "resource" to use and is an optional argument
- This happens in [getActiveDirectoryResourceId()](https://github.com/Azure/ms-rest-nodeauth/blob/3710a14c26a584dd667a0f726fe015f868d8cdcd/lib/credentials/tokenCredentialsBase.ts#L59) where you can see that the default value that gets used is Environment.activeDirectoryResourceId which for public cloud is `https://management.core.windows.net/` and for china cloud is `https://management.core.chinacloudapi.cn/`
- So, am confused why things are working when you pass `https://management.azure.com/` as the tokenAudience which is different from the default value that works for all mgmt plane packages.
- Also, from my understanding, a user only ever had to pass tokenAudience if they were using a data plane library

What I am trying to get at is whether we need to show the use of tokenAudience in this code snippet at all.

I then realized that you have to do that because for the China Cloud example, you need to pass the environment, but tokenAudience is the optional parameter that appears in the constructor before environment.

Which then makes me wonder if we should use the `withServicePrincipalSecret()` method here which uses an options bag instead of `new ApplciationTokenCredentials`

sadasant replied:
Ohh I see now! I’ll take your advice, I will use `withServicePrincipalSecret()`.

sadasant replied:
Alright, updated! How does it look now?

ramya-rao-a commented on migrate-to-identity-v2.md:
```diff
@@ -221,7 +221,8 @@ async function main() {
 -  });
-+  const credential = new AzureCliCredential("https://vault.azure.net/.default");
++  const credential = new InteractiveBrowserCredential();
++  const accessToken = await credential.getToken("https://vault.azure.net/.default");
```
If we are showing how to get the access token using azure/identity, shouldnt we also show the same for ms-rest-nodeauth here?

sadasant replied:
How does it look now? interactiveLogin returns the device code credential (in plural `credentials` is more commonly called in `ms-rest-nodeauth`). That credential then has a similar getToken method, but one that does not receive parameters.

sadasant commented:
I’ll follow up tomorrow!

sadasant committed:
Update migrate-to-identity-v2.md

A more relevant token audience on the sample for national clouds, and using getToken on the getToken section about scopes

sadasant committed:
Using loginWithServicePrincipalSecret

ramya-rao-a commented on migrate-to-identity-v2.md:
```diff
@@ -119,17 +119,14 @@ import * as dotenv from "dotenv";
++ const credential = new ClientSecretCredential(domain, clientId, secret, { authorityHost });
   const client = new SubscriptionClient(credential, {
 -  baseUri: environment.resourceManagerEndpointUrl,
 +  baseUri: "https://management.chinacloudapi.cn"
```
Do we need to hard code this? Can we not fetch the value from `environment`?

sadasant replied:
I’ll do it! — just feels kinda weird. We don’t generally recommend using ms-rest-azure-env with the new Identity package. On my way!

sadasant replied:
Done!

ramya-rao-a replied:
That is true, I didnt realize that when I recommended using `environment`.

I would be ok if we want to go back to hard-coding, but this time pull the value of `resourceManagerEndpointUrl` out into a constant

sadasant replied:
For migration purposes, we don’t have a way for people to get this baseUri with the new libraries. They generally wouldn’t need to do this. I’m wondering if we should keep ms-rest-env as a useful tool only within this migration guide.

sadasant replied:
I think I should merge this for now. We can revisit this later.

sadasant committed:
Reusing environment.resourceManagerEndpointUrl

ramya-rao-a commented on migrate-to-identity-v2.md:
```diff
@@ -108,9 +108,9 @@ const credential = new ClientSecretCredential(
+- import { loginWithServicePrincipalSecret } from "@azure/ms-rest-nodeauth";
 + import { ClientSecretCredential, AzureAuthorityHosts } from "@azure/identity";
-- import { Environment } from "@azure/ms-rest-azure-env";
+import { Environment } from "@azure/ms-rest-azure-env";
```
`Environment` is not needed when using `@azure/identity`
```suggestion
- import { Environment } from "@azure/ms-rest-azure-env";
```

sadasant replied:
I’m using it to set the `baseUri: environment.resourceManagerEndpointUrl,`, in order to not have the raw value.

ramya-rao-a replied:
Hmm... that might be problematic because eventually we want to deprecate all the `@azure/ms-rest*` packages. But thats a problem for another day :)

ramya-rao-a commented on migrate-to-identity-v2.md:
```diff
@@ -119,20 +119,16 @@ import * as dotenv from "dotenv";
-- const environment = Environment.ChinaCloud;
+const domain = process.env.AZURE_TENANT_ID; // domain or tenantId
+const secret = process.env.AZURE_CLIENT_SECRET;
+const environment = Environment.ChinaCloud;
```
`environment` is not needed when using `@azure/identity`
```suggestion
- const environment = Environment.ChinaCloud;
```

sadasant replied:
I’m using it to set the `baseUri: environment.resourceManagerEndpointUrl,`, in order to not have the raw value.
